contracts/exchange/contracts/src/MixinTransactions.sol

```diff
@@ -46,44 +48,51 @@ contract MixinTransactions is
     )
         public
     {
-        // Prevent reentrancy
-        require(
-            currentContextAddress == address(0),
-            "REENTRANCY_ILLEGAL"
-        );
-
         bytes32 transactionHash = getTransactionHash(transaction);
 
+        // Prevent reentrancy
+        if (currentContextAddress != address(0)) {
+            rrevert(TransactionError(
+                transactionHash,
+                TransactionErrorCodes.NO_REENTRANCY
+            ));
+        }
+
         // Validate transaction has not been executed
-        require(
-            !transactions[transactionHash],
-            "INVALID_TX_HASH"
-        );
+        if (transactions[transactionHash]) {
+            rrevert(TransactionError(
+                transactionHash,
+                TransactionErrorCodes.ALREADY_EXECUTED
+            ));
+        }
 
         // Transaction always valid if signer is sender of transaction
         address signerAddress = transaction.signerAddress;
         if (signerAddress != msg.sender) {
             // Validate signature
-            require(
-                isValidSignature(
+            if (!isValidSignature(
                     transactionHash,
                     signerAddress,
-                    signature
-                ),
-                "INVALID_TX_SIGNATURE"
-            );
+                    signature)) {
+                rrevert(TransactionError(
```

Comment:
The validation function could just work on a copy of the signature. That's probably cleaner from a design standpoint since it isn't obvious that the validation function has side effects in the first place. If you agree with this approach, I'll make the change in #1774 after the rebase since that PR is all up in the validation function's business.

Reply:
We don't copy the signature for efficiency reasons (and there isn't any other place in the contracts where it would even matter that the original signature is being mutated). A couple of other options:

- Copy the `signatureType` byte and concat it with the signature after validating (only when we need to reuse the signature after validating). Pretty dirty, but more will probably be the most efficient in the long run.
- In the validation function, use the last byte of the signature as the `signatureType` but do _not_ pop it off. If we do need to pass the signature into an external function call (only the `Wallet` or `Validator` types), then we copy the signature and pop off the last byte from the copy. This seems like probably the cleanest approach.

Reply:
Second sounds great.

Reply:
[I went ahead and did it in this PR](https://github.com/0xProject/0x-monorepo/blob/9c104f4e88b8d7dc69fc1d461a414a79865c2de2/contracts/exchange/contracts/src/MixinSignatureValidator.sol#L305).

For `SignatureType.Wallet` and `SignatureType.Validator` I took a hacky approach by just resizing the signature before encoding the call data then restoring it immediately afterwards:

```solidity
// Shave the validator address and signature type from the signature.
assembly {
    mstore(signature, sub(signatureLength, 21))
}
// Encode the call data.
bytes memory callData = abi.encodeWithSelector(
    IValidator(validatorAddress).isValidSignature.selector,
    hash,
    signerAddress,
    signature
);
// Restore the full signature.
assembly {
    mstore(signature, signatureLength)
}
// Static call the verification function.
(bool didSucceed, bytes memory returnData) = validatorAddress.staticcall(callData);
```
I think this is safe, but let me know if you see any issues.

Reply:
Nice, I like this approach.